service/udp.go

```diff
@@ -79,130 +84,229 @@ func findAccessKeyUDP(clientIP netip.Addr, dst, src []byte, cipherList CipherLis
+			}
+			return &timedPacketConn{
+				PacketConn:     pc,
+				defaultTimeout: natTimeout,
+			}, nil
+		},
 	}
 }
 
-// PacketHandler is a running UDP shadowsocks proxy that can be stopped.
-type PacketHandler interface {
+// AssociationHandler is a handler that handles UDP assocations.
+type AssociationHandler interface {
+	Handle(association net.Conn, metrics UDPAssocationMetrics)
 	// SetLogger sets the logger used to log messages. Uses a no-op logger if nil.
 	SetLogger(l *slog.Logger)
 	// SetTargetIPValidator sets the function to be used to validate the target IP addresses.
 	SetTargetIPValidator(targetIPValidator onet.TargetIPValidator)
-	// Handle returns after clientConn closes and all the sub goroutines return.
-	Handle(clientConn net.PacketConn)
+	// SetTargetConnFactory sets the function to be used to create new target connections.
+	SetTargetConnFactory(factory func() (net.PacketConn, error))
 }
 
-func (h *packetHandler) SetLogger(l *slog.Logger) {
+func (h *associationHandler) SetLogger(l *slog.Logger) {
 	if l == nil {
 		l = noopLogger()
 	}
 	h.logger = l
 }
 
-func (h *packetHandler) SetTargetIPValidator(targetIPValidator onet.TargetIPValidator) {
+func (h *associationHandler) SetTargetIPValidator(targetIPValidator onet.TargetIPValidator) {
 	h.targetIPValidator = targetIPValidator
 }
 
-// Listen on addr for encrypted packets and basically do UDP NAT.
-// We take the ciphers as a pointer because it gets replaced on config updates.
-func (h *packetHandler) Handle(clientConn net.PacketConn) {
-	nm := newNATmap(h.natTimeout, h.m, h.logger)
-	defer nm.Close()
-	cipherBuf := make([]byte, serverUDPBufferSize)
-	textBuf := make([]byte, serverUDPBufferSize)
+func (h *associationHandler) SetTargetConnFactory(factory func() (net.PacketConn, error)) {
+	h.targetConnFactory = factory
+}
 
+type AssocationHandleFunc func(assocation net.Conn)
+
+// PacketServe listens for UDP packets on the provided [net.PacketConn], creates
+// creates and manages NAT associations, and invokes the provided `handle`
+// function for each association. It uses a NAT map to track active associations
+// and handles their lifecycle.
+func PacketServe(clientConn net.PacketConn, handle AssocationHandleFunc, metrics NATMetrics) {
+	nm := newNATmap()
+	defer nm.Close()
+	buffer := make([]byte, serverUDPBufferSize)
 	for {
-		clientProxyBytes, clientAddr, err := clientConn.ReadFrom(cipherBuf)
-		if errors.Is(err, net.ErrClosed) {
-			break
+		n, addr, err := clientConn.ReadFrom(buffer)
+		if err != nil {
+			if errors.Is(err, net.ErrClosed) {
+				break
+			}
+			slog.Warn("Failed to read from client. Continuing to listen.", "err", err)
+			continue
 		}
+		pkt := buffer[:n]
+
+		// TODO: Include server address in the NAT key as well.
+		conn := nm.Get(addr.String())
+		if conn == nil {
+			conn = &natconn{
+				Conn:   &packetConnWrapper{PacketConn: clientConn, raddr: addr},
+				readBufCh: make(chan []byte, 1),
+				bytesReadCh: make(chan int, 1),
+			}
```

Comment:
I think you can get rid of `packetConnWrapper`. Just put the address in the `natconn`.

```suggestion
			conn = &natconn{
				Conn:   clientConn,
				raddr: addr,
				readBufCh: make(chan []byte, 1),
				bytesReadCh: make(chan int, 1),
			}
```